### PR TITLE
Adding WinMD generator

### PR DESCRIPTION
**Pull request by manodasanW**
Adding a source generator which parses a C# project and generates a WinMD with the public types from it.  The WinMD generator isn't complete and there is some missing scenarios in it currently, but it generates a WinMD for most of the basic scenarios.  Portions of the WinMD generation code can be simplified to use the functions which generate WinMD types from the symbol  throughout but that would be addressed in a future PR.

Also adding a sample to test the source generator.

Contributes to #309

### CODE REVIEW

File: Authoring/AuthoringSample/AuthoringSample.csproj [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+    <LangVersion>preview</LangVersion>
+    <AssemblyVersion>1.0.0.0</AssemblyVersion>
+    <EnableCsWinRTLogging>true</EnableCsWinRTLogging>
```
Comment (Scottj1s) [inline Authoring/AuthoringSample/AuthoringSample.csproj:7]: do you expect to make this permanent?  Can we rename to CsWinRTEnableLogging?

File: Authoring/AuthoringSample/AuthoringSample.csproj [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+    <LangVersion>preview</LangVersion>
+    <AssemblyVersion>1.0.0.0</AssemblyVersion>
+    <EnableCsWinRTLogging>true</EnableCsWinRTLogging>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <CompilerVisibleProperty Include="AssemblyName" />
```
Comment (Scottj1s) [inline Authoring/AuthoringSample/AuthoringSample.csproj:11]: I assume AssemblyName has a reasonable default value?

File: Authoring/AuthoringSample/AuthoringSample.csproj (hunk 1 above)
Comment (manodasanW) [reply Authoring/AuthoringSample/AuthoringSample.csproj:7]: Yea I think we are going to need some form of logging mode.  I can rename it.

File: Authoring/WinRT.SourceGenerator/Generator.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+        private string GetAssemblyVersion(SourceGeneratorContext context)
+        {
+            context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.AssemblyVersion", out var assemblyVersion);
+            return assemblyVersion;
+        }
+
+        private string GetWinmdOutputFile(SourceGeneratorContext context)
+        {
+            string winmdDir = Path.Combine(Directory.GetCurrentDirectory(), "Generated Files");
```
Comment (Scottj1s) [inline Authoring/WinRT.SourceGenerator/Generator.cs:27]: GeneratedFiles can be defaulted like this, but should also support a project definition (as cppwinrt does)

Comment (Scottj1s) [reply Authoring/WinRT.SourceGenerator/Generator.cs:27]: One reason for that is cswinrt can generate different code based on the -tpm switch, in which case we'd want to generated sources under $IntDir

File: Authoring/WinRT.SourceGenerator/Generator.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            Directory.CreateDirectory(winmdDir);
+            return Path.Combine(winmdDir, GetAssemblyName(context) + ".winmd");
+        }
+
+        private void GenerateWinMD(MetadataBuilder metadataBuilder, string outputFile)
+        {
+            Logger.Log("Writing " + outputFile);
+            var managedPeBuilder = new ManagedPEBuilder(
+                new PEHeaderBuilder(imageCharacteristics: Characteristics.ExecutableImage),
```
Comment (Scottj1s) [inline Authoring/WinRT.SourceGenerator/Generator.cs:36]: you don't have to OR in Characteristics.Dll ?

File: Authoring/WinRT.SourceGenerator/Logger.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+namespace Generator
+{
+    class Logger
+    {
+        public static void Initialize(SourceGeneratorContext context)
+        {
+            context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.EnableCsWinRTLogging", out var enableLogging);
+            if(enableLogging != null && bool.Parse(enableLogging))
+            {
+                string generatedDir = Path.Combine(Directory.GetCurrentDirectory(), "Generated Files");
```
Comment (Scottj1s) [inline Authoring/WinRT.SourceGenerator/Logger.cs:13]: this should be factored into a shared function, GetGeneratedFiles

File: Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs [hunk 6]
```diff
@@ -0,0 +1,1791 @@
...
+        }
+    }
+
+    class TypeDeclaration
+    {
+        public readonly ISymbol Node;
+        public TypeDefinitionHandle Handle;
+        public string DefaultInterface;
+
+        public Dictionary<ISymbol, List<MethodDefinitionHandle>> MethodDefinitions;
```
Comment (Scottj1s) [inline Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs:73]: can inline assign these

File: Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs [hunk 7]
```diff
@@ -0,0 +1,1791 @@
...
+            {
+                MethodReferences[node] = new List<EntityHandle>();
+            }
+
+            MethodReferences[node].Add(handle);
+        }
+
+        public List<MethodDefinitionHandle> GetMethodDefinitions()
+        {
+            return MethodDefinitions.Values.SelectMany(list => list).ToList();
```
Comment (Scottj1s) [inline Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs:118]: can't just use MethodDefinitions.Values.ToList() ?

File: Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs [hunk 8]
```diff
@@ -0,0 +1,1791 @@
...
+        }
+
+        public List<EntityHandle> GetMethodReferences()
+        {
+            return MethodReferences.Values.SelectMany(list => list).ToList();
```
Comment (Scottj1s) [inline Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs:123]: same as above

File: Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs (hunk 7 above)
Comment (manodasanW) [reply Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs:118]: No, that makes it a list of lists.  I am trying to merge all the method definitions created for each symbol (i.e. property, method, event, etc.) into 1 list.

File: Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs [hunk 9]
```diff
@@ -0,0 +1,1791 @@
...
+                isInterfaceParent,
+                method.MethodKind == MethodKind.Constructor,
+                true,
+                true);
+            currentTypeDeclaration.AddMethod(method, methodDefinitionHandle);
+        }
+
+        void AddFactoryMethod(INamedTypeSymbol classSymbol, IMethodSymbol method)
+        {
+            Logger.Log("ading factory method: " + method.Name);
```
Comment (Scottj1s) [inline Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs:1305]: typo

File: Authoring/AuthoringSample/AuthoringSample.csproj (hunk 2 above)
Comment (manodasanW) [reply Authoring/AuthoringSample/AuthoringSample.csproj:11]: Yes in this sample the default seems to be AuthoringSample.

File: Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs [hunk 10]
```diff
@@ -0,0 +1,1791 @@
...
+            AddSynthesizedInterface(
+                classDeclaration,
+                SynthesizedInterfaceType.Factory,
+                classMembersFromInterfaces);
+
+            AddSynthesizedInterface(
+                classDeclaration,
+                SynthesizedInterfaceType.Default,
+                classMembersFromInterfaces);
+        }
```
Comment (Scottj1s) [inline Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs:1554]: overridable and composable?

Comment (manodasanW) [reply Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs:1554]: on the todo, will be added once I got the cswinrt side working for basic components.  Will add a todo comment.

File: Authoring/WinRT.SourceGenerator/Generator.cs (hunk 3 above)
Comment (manodasanW) [reply Authoring/WinRT.SourceGenerator/Generator.cs:27]: Agree on this, going to defer implementing this until we figure out the right place to write to.

File: Authoring/WinRT.SourceGenerator/Generator.cs (hunk 4 above)
Comment (manodasanW) [reply Authoring/WinRT.SourceGenerator/Generator.cs]: Updated and added a few more flags.

**Pull request merged**